Add org.pwsafe.pwsafe

ronys opened the pull request:
### Please confirm your submission meets all the criteria

- [y] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [y] My pull request follows the instructions at [App Submission][submission].
- [y] I have [built][build] and tested the submission locally.
- [y] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
- [y] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [y] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. **Link:**
- [y] The domain used for the application ID is controlled by the application developers either directly or through the code hosting (e.g. GitHub, GitLab, SourceForge, etc.). The [application id guidelines][app-id] are followed.
- [y] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*

[reqs]: https://docs.flathub.org/docs/for-app-authors/requirements
[maint]: https://docs.flathub.org/docs/for-app-authors/maintanance
[submission]: https://docs.flathub.org/docs/for-app-authors/submission
[build]: https://docs.flathub.org/docs/for-app-authors/submission/#before-submission
[app-id]: https://docs.flathub.org/docs/for-app-authors/requirements#application-id

hfiguiere commented on pwsafe64x64.png:
you don't need that icon size. it's also never referenced anywhere.

igor2x replied:
From the link I have submitted to application developer, I now see there is broken ling for flapak/flathub Requirements. I am sure four months ago there was a requirement to submit two icon files 64x64 and 128x128 icons. I am sure, because I had to created them manually, struggling not to create a mess out of icons.

If this file is not required anymore, then @ronys, can you please remove it from repository.

hfiguiere commented on pwsafe128x128.png:
this file is never referenced anywhere. 

there is no icon in the upstream project?

igor2x replied:
Four months ago I have created this 128x128 file, because there was a requirement that program icon should have some margin. Original icon is created to the very left and bottom of the icon border, no margin. I have "fixed" the icon to move icon from the left and bottom to the middle and was required to change the whole icon little bit, to follow the logic of original icon. I can't see any Flatpak/Flathub requirement about icon "margin" anymore. Looks like this requirement was removed or something...

@ronys, can you look at this icon and see if it is better than original one, and if yes, please commit it to Password Safe repository.

hfiguiere commented on org.pwsafe.pwsafe.xml:
this file isn't referenced nor installed anywhere.

hfiguiere replied:
this is the source of the build error.

bbhtt replied:
the extension should be appdata.xml or metainfo.xml not just xml see https://docs.flatpak.org/en/latest/conventions.html#appdata-files

It should also live in the upstream repository.

igor2x replied:
@ronys, this looks like also changed from flatpak/flatpak Requirement documentation.

hfiguiere commented:
bot, build org.pwsafe.pwsafe

bbhtt commented on org.pwsafe.pwsafe.xml:
```diff
@@ -0,0 +1,33 @@
+  </releases>
+  <screenshots>
+    <screenshot type="default">
+      <image>https://www.pwsafe.org/images/treeview.png</image>
```
Can there be a bigger screenshot and one that is on Linux?

igor2x replied:
@ronys, I have created print screen from Ubuntu 23.10, but it should look the same on any Linux distro. See if you like the picture and upload it like to www.pwsafe.org web site.

![PasswordSafe_on_Linux](https://github.com/flathub/flathub/assets/10895030/f53d6643-b519-4705-8563-c1a48420243e)

I just created few "groups" and "entries" and added right menu to display how easy is to add new entry.

ADDITIONALLY:
@ronys, I think there should be written somewhere that Password Safe is cross platform: Windows, Linux and MacOS. This is in my humble opinion the main strength of the program. Few years ago, when I was searching for password manager tool, this was the only decent cross-platform tool I have found (because I use Windows at my daily job and Ubuntu at home).

igor2x commented on org.pwsafe.pwsafe.yml:
```diff
@@ -0,0 +1,65 @@
+  - --env=PWS_HELPDIR=/app/share/passwordsafe/help/
+  # YubiKey USB access
+  - --device=all
+cleanup-commands:
```
I am user of Password Safe program and few months ago I have created manifest file, build flatpak successfully, zipped all the files required and uploaded them to [application developer github issue tracker site](https://github.com/pwsafe/pwsafe/issues/562) and finally communicated with developer, he will submit the flatpak. Developer @ronys was busy and now after few months reported submitted manifest...

I have only built one single flatpak, this one, and like I remember there were two commands to clean up. First one was for deleting files and I was struggling with it to delete whole subdirectories, so that is the reason I used "cleanup-commands" option, to just use "rm" Linux command to delete all required subdirectories. Any help how to clean those subdirectories is much appreciated.

hfiguiere replied:
https://docs.flatpak.org/en/latest/manifests.html#cleanup

ronys commented:
Thanks for the review comments. I've updated the PR and added the metainfo xml file to the pwsafe repo and build process.
Note, however, that this file will only be installed in releases after the current tag (1.18.2). Does this mean we need to define a new tag/release for flatpak?

bbhtt commented on org.pwsafe.pwsafe.yml:
```diff
@@ -0,0 +1,62 @@
+        tag: v3.2.4
+        commit: 5052c90b067dcc347d58822b450897d16e2c31e5
+    cleanup:
+      - '/app/lib/cmake/XercesC/*.cmake'
```
```suggestion
      - /lib/cmake
```

bbhtt commented:
> Thanks for the review comments. I've updated the PR and added the metainfo xml file to the pwsafe repo and build process. Note, however, that this file will only be installed in releases after the current tag (1.18.2). Does this mean we need to define a new tag/release for flatpak?

They are required for the build, please make a new tag with them and update the tag/commit of pwsafe here.

bbhtt commented:
bot, build org.pwsafe.pwsafe

igor2x commented:
@ronys, build error: _"Error: Error opening file /srv/buildbot/worker/build-x86_64-7/build/.flatpak-builder/rofiles/rofiles-lP8JdF/files/share/appdata/org.pwsafe.pwsafe.appdata.xml: Is a directory"_

Looking at commit: https://github.com/pwsafe/pwsafe/commit/726afaacc8fd55a75fd9b984cf4988a34434362f there is line:
`install (FILES "install/metainfo/org.pwsafe.pwsafe.metainfo.xml" DESTINATION "share/metainfo/org.pwsafe.pwsafe.metainfo.xml")`

But parameter DESTINATION should point to directory not to file. Probably correct is to point just to directory without file name at the end.

`install (FILES "install/metainfo/org.pwsafe.pwsafe.metainfo.xml" DESTINATION "share/metainfo/")`

ronys commented:
Fixed the install destination. Hope we're good this time.

bbhtt commented:
bot, build org.pwsafe.pwsafe

igor2x commented:
@ronys, from error message:

    "errors": [
        "appstream-failed-validation",
        "appstream-missing-icon-file",
        "appstream-missing-developer-name"
    ],
    "warnings": [
        "appstream-screenshot-missing-caption",
        "appstream-summary-too-long"
    ],
    "appstream": [
        "tag-missing           : <release> required",
        "tag-missing           : <url> is not present",
        "tag-missing           : <content_rating> required [use https://odrs.gnome.org/oars]"
    ]
}

1. I have fixed all errors/warnings except "appstream-missing-icon-file" (see second point bellow). Please replace https://github.com/ronys/flathub/blob/pwsafe/org.pwsafe.pwsafe.yml file with file inside zip file: [metainfo.zip](https://github.com/flathub/flathub/files/14099361/metainfo.zip)

2. Error "appstream-missing-icon-file" must be fixed in building. Please see instructions how to fix this at: https://docs.flathub.org/docs/for-app-authors/linter/#appstream-missing-icon-file

igor2x commented:
@ron I am investigating the second problem  "appstream-missing-icon-file"

After flatpak is created on my Ubuntu computer I can enter inside flatpak sandbox with:
`flatpak enter org.pwsafe.pwsafe bash`

Then according to [icons instructions](https://docs.flatpak.org/en/latest/conventions.html#appdata-files) the path of icon should be:
/app/share/icons/hicolor/128x128/apps/org.pwsafe.pwsafe.png 

but there is no such above path inside Password Safe flatpak sandox. Check with:
ls /app/share/icons/hicolor/
and only "48x48" directory is present.

Like I understand, icon get automatically created (copied) during flatpak creation from flatpak upstream project.
In upstream project I have found the following icon:
https://github.com/pwsafe/pwsafe/blob/master/install/graphics/pwsafe.png
If I open file with image editor it is clear this icon is 48x48 in size.

I assume all you have to do is to replace https://github.com/pwsafe/pwsafe/blob/master/install/graphics/pwsafe.png file with for example 128x128 PNG file I have created (and you have committed in first commit of this repository) or maybe simpler to just download https://github.com/pwsafe/pwsafe/files/12566476/flathub.zip file, unzip file and there is file with name pwsafe128x128.png. Rename this file to pwsafe.png and copy it to https://github.com/pwsafe/pwsafe/blob/master/install/graphics/pwsafe.png

Note: This is probably one of the last things we have to do. We see Flathub is pretty mature project, because it requires all sort of quality controls. This may be annoying, but it must be done only for the first flatpak build.

igor2x commented:
@bbhtt, I review the changes and it looks good to me. Can you build please?

bbhtt commented:
bot, build org.pwsafe.pwsafe

Feel free to trigger builds when you need.

igor2x commented:
@ronys, ahhhh now I see you have not performed the tasks from my first post first point where I have updated all of the required fields.

    "errors": [
        "appstream-missing-developer-name"
    ],
    "warnings": [
        "appstream-summary-too-long",
        "appstream-screenshot-missing-caption"
    ]

OK, lets get step by step, to create official review. All the errors are from org.pwsafe.pwsafe.metainfo.xml file.
1. Error: "appstream-missing-developer-name". Info: https://docs.flathub.org/docs/for-app-authors/linter/
After existing line in file
`<project_license>Artistic-2.0</project_license>`
add new line:
`<developer_name>Rony Shapiro</developer_name> `

2. Warning: "appstream-summary-too-long". Info: https://docs.flathub.org/docs/for-app-authors/linter/
Summary should not exceed 35 characters.
Search for the line:
`<summary>Password Safe is a password management utility.</summary>`
and change it to:
`<summary>Password management utility</summary>`
There is no need to use "Password Safe" in this field, because program name is already there. This is just simple summary.
If I look go the Gnome Software where Password Safe program is displayed (because I have installed it from local flatpak package): 
![image](https://github.com/flathub/flathub/assets/10895030/71f938ea-ec79-4396-b392-541d39e7b308)
You see removing "Password Safe" from the string is the best option to not repeat your self.

3. Warning: "appstream-screenshot-missing-caption". Info: https://docs.flathub.org/docs/for-app-authors/linter/
```
 <screenshots>                                                
   <screenshot type="default">                                
     <caption>Sample tree view"</caption>                     
     <image>https://pwsafe.org/images/wxTreeview.png</image>  
   </screenshot>                                              
 </screenshots>       
```
Note: Be careful your string ends with `"` character. You see: `view"`

Also (I don't know if this is requirement, but I have seen the following order in other projects xml file), so move the "caption" tag one row bellow the "image" tag to get:
```
 <screenshots>                                                
   <screenshot type="default">                                
     <image>https://pwsafe.org/images/wxTreeview.png</image>  
     <caption>Sample tree view</caption>                     
   </screenshot>                                              
 </screenshots>       
```
Note: Make sure there is not additional `"` character in text.

Like I understand @bbhtt you can start a build your self when you are ready with command from https://github.com/flathub/flathub/pull/4890#issuecomment-1915918676

igor2x commented:
Now I have noticed that from yesterday error message:
`"tag-missing           : <url> is not present",`
you have only added one setting:
`<url type="homepage">`
which is fine and is bare minimum, but bellow above line please also add this new lines, to provide info for end-users:
`<url type="help">https://pwsafe.org/quickstart.shtml</url>`
`<url type="bugtracker">https://github.com/pwsafe/pwsafe/issues</url>`
`<url type="translate">https://explore.transifex.com/passwordsafe/passwordsafe/</url>`
`<url type="faq">https://pwsafe.org/faq.shtml</url>`

Those are just additional info end-user can get from Flathub repository and operating system stores. For example for Firefox this info looks like bellow on Flathub:
![image](https://github.com/flathub/flathub/assets/10895030/8347739f-80c0-4fa8-a517-eec98c9288fe)

Info from:  https://docs.flathub.org/docs/for-app-authors/appdata-guidelines/#url

ronys commented:
bot, build org.pwsafe.pwsafe

igor2x commented:
@ronys, I have run above command and double "y" to confirm and then run: `flatpak run org.pwsafe.pwsafe` to start app. I tested the flatpak and it works fine, just like locally build one. Great job!!!

@bbhtt, is there anything else we should do? I tried searching the Flathub for "Password Safe" with https://flathub.org/apps/search?q=Password%20Safe and application is not found. Maybe there is some refresh frequency we need to wait to.

ronys commented:
@igor2x - thanks for your help. Without your assistance I would have given up long ago.

BTW - I noticed your comment re the url tqgs only after the last commit. Will add this for the next version.

igor2x commented:
Like I understand the process:
1. Now flatpak is built successfully and it got propagated to the "test server".
2. Testers (like you @ronys and me) install flatpak from "test server" and test the flatpak.
3. If testing is successful, then I assume must be some propagation of flatpak package from "test" to "production" server
   and end-users can install the package from official production Flathub.

@ronys do you know how to do the from test to production propagation? If not, maybe @bbhtt can help us to do, I assume one last step.

ronys commented:
According to the [documentation](https://docs.flathub.org/docs/for-app-authors/submission/):
"Once your pull request has been approved, a new repository will be created for your application, containing your app’s Flatpak manifest. You will receive an invitation to have write access for the repository, please make sure to accept it before it expires."

So we're waiting for this stage.

ronys commented:
Any update?

barthalion commented:
bot, build org.pwsafe.pwsafe
